src/app_builder.rs

```diff
@@ -75,164 +247,166 @@ impl AppBuilder {
+        if self
+            .signature
+            .track_update_packed
+            .associate_plugin::<T>(&self.track_current_plugin, reason)
+            .is_first()
+        {
+            self.app.world.borrow::<ViewMut<T>>().unwrap().update_pack();
+            self.resets.push(system!(reset_update_pack::<T>));
+        }
 
-        match self.track_update_packed.entry(type_id) {
-            Entry::Occupied(mut list) => {
-                // no need to pack again
-                list.get_mut()
-                    .push((self.track_current_plugin.clone(), reason));
-            }
-            Entry::Vacant(list) => {
-                list.insert(vec![(self.track_current_plugin.clone(), reason)]);
-                self.world.borrow::<ViewMut<T>>().update_pack();
-                self.add_systems_to_stage(stage::LAST, |workload| {
-                    workload.with_system(system!(reset_update_pack::<T>));
-                });
-            }
+        self
+    }
+
+    /// Declare dependency on `[crate::Tracked]<T>` and add "reset tracked" as the last system.
+    #[track_caller]
+    pub fn tracks<T: 'static + Send + Sync>(&mut self, reason: &'static str) -> &mut Self {
+        if self
+            .signature
+            .track_tracked_uniques
+            .associate_plugin::<T>(&self.track_current_plugin, reason)
+            .is_first()
+        {
+            self.resets.push(system!(reset_tracked_unique::<T>));
         }
 
         self
     }
 
     /// Add a unique component
+    #[track_caller]
     pub fn add_unique<T>(&mut self, component: T) -> &mut Self
     where
         T: Send + Sync + 'static,
     {
-        self.world.add_unique(component);
-        let unique_type_id = self.tracked_type_id_of::<T>();
-        self.track_uniques
-            .entry(unique_type_id)
-            .or_default()
-            .push(self.track_current_plugin.clone());
+        if self
+            .signature
+            .track_uniques_provided
+            .associate_plugin::<T>(&self.track_current_plugin, "<not provided>")
+            .is_first()
+        {
+            self.app.world.add_unique(component).unwrap();
+        } else {
+            warn!(
+                "Unique({}) already provided by another plugin in the plugin workload.",
+                type_name::<T>(),
+            )
+        }
+
+        self
+    }
+
+    /// Add a tracked unique value.
+    ///
+    /// Accessible through the [crate::Tracked] and [crate::TrackedMut] views.
+    #[track_caller]
+    pub fn add_tracked_value<T>(&mut self, component: T) -> &mut Self
+    where
+        T: Send + Sync + 'static,
+    {
+        if self
+            .signature
+            .track_tracked_uniques_provided
+            .associate_plugin::<T>(&self.track_current_plugin, "<not provided>")
+            .is_first()
+        {
+            self.app
+                .world
+                .add_unique(TrackedValue::new(component))
+                .unwrap();
```

Comment:
Again, a comment explaining why unwrapping is OK here would be nice.

Reply:
Would a panic section note in the doc comment suffice?

maybe `expect("add_tracked_value has exclusive access to borrowing this storage")`?